### DISCUSSION TITLE
Help needed to make helloSystem Ventoy Compatible

### DISCUSSION THREAD

**Discussion by probonopd** [ventoy/Ventoy#1235] [Q&A] [answered]
[helloSystem](https://hellosystem.github.io/) is a FreeBSD based desktop operating system Live ISO.

The early boot phase of the Live ISO is very similar to GhostBSD, but unlike GhostBSD, the latest [experimental helloSystem 0.7.0 ISOs](https://github.com/helloSystem/ISO/releases/tag/experimental-13.0) (e.g., `hello-0.7.0_0G48-FreeBSD-13.0-amd64.iso`) currently cannot be booted with Ventoy. (Earlier versions were more like FuryBSD and worked.)

I would like to make it fully Ventoy Compatible, but even after a lot of experimentation I could not find out why Ventoy can boot GhostBSD but not helloSystem.

I suspect the answer lies deep down in https://github.com/ventoy/Ventoy/tree/master/Unix/ventoy_unix_src/FreeBSD/geom_ventoy_src but it seems that it is entirely undocumented. From what I can see it relies on certain files (e.g., `COPYRIGHT`) being in certain locations and such.

**Please help me make helloSystem fully Ventoy Compatible.**

Ideally not by having helloSystem specific workaround hardcoded in Ventoy, but by me changing whatever needs to be changed on the helloSystem ISO so that it can be booted with Ventoy just like https://download.freebsd.org/ftp/releases/ISO-IMAGES/13.0/FreeBSD-13.0-RELEASE-amd64-bootonly.iso or GhostBSD.

Thanks!

**Comment by ventoy** [comment]
`geom_ventoy` is a driver for FreeBSD. So if you can build the source and include the geom_ventoy.ko in your distro and load it default in bootloader, then is can be Ventoy compatible.

**Reply by probonopd** [reply]
https://download.freebsd.org/ftp/releases/ISO-IMAGES/13.0/FreeBSD-13.0-RELEASE-amd64-bootonly.iso or GhostBSD seemingly don't do this, but still work.

Is including the driver in the distribution and preloading it by the bootloader the preferred (less hackish) more robust way? I can do that. In fact I had tried that but it did not work. I will try again.

Thanks for your help.

**Reply by ventoy** [reply]
Ventoy must prebuild the ko and do some hook for FreeBSD.
The hook is to insert the ko to standard FreeBSD iso and make it loaded by the bootloader.
But the ko is version dependency, so if FreeBSD 14.0/15.0... release, Ventoy must update to support them.

So if you include the ko in your system then Ventoy will don't need to do the  hook and it's more compatible.

Besides, you must mark the ISO as follows, then Ventoy will known that the ISO contains compatible process for Ventoy, then Ventoy will not try to do the hook anymore.
https://www.ventoy.net/en/doc_compatible_mark.html

**Reply by probonopd** [reply]
I have built a FreeBSD 13.0-RELEASE based [ISO of helloSystem](https://github.com/helloSystem/ISO/releases/tag/experimental-13.0) that includes and loads the Ventoy kernel module.

When I boot with `-v`, then I can see `g_ventoy_taste(VENTOY, da0)`

but never

```
######## ventoy disk <da0> ########
```

I think the error is somewhere around here:

![image](https://user-images.githubusercontent.com/2480569/141674428-71cb23a0-2e6e-434e-96bf-3b82683c08c5.png)

When I boot GhostBSD ISO from **the same Ventoy installation on the same Ventoy disk (exfat)**, then I _do_ get `######## ventoy disk <da0> ########`.

What is the next step in debugging?

Thanks for your help.

**Comment by ventoy** [comment]
Sorry.
My bad. I forgot that only load the ko is not enough. Ventoy still need to hook to write some data to the load.conf, the data is used in the geom_ventoy.ko.
But maybe Ventoy can work with helloSystem to make it more easy to boot from Ventoy.

For standard FreeBSD, Ventoy need to pack the prebuild ko and load it when boot and write the data to load.conf. But the ko need to update with the new FreeBSD relase.

helloSystem  can include the ok and auto load it in load.conf. Then Ventoy will only need to write the data to load.conf to boot it.
When helloSystem update, Ventoy don't need to update.

**Reply by probonopd** [reply]
Thanks for your help.

What do I need to do exactly? What data do I need to add to load.conf? Is it documented somewhere?

Or does Ventoy need to do something?

It is important for me that it will work for future helloSystem ISOs too (we build several each week), without waiting for changes in Ventoy each time we make a new ISO.

**Comment by ventoy** [comment]
1. Use the latest https://github.com/ventoy/Ventoy/tree/master/Unix/ventoy_unix_src/FreeBSD/geom_ventoy_src/13.x/sys  source for the driver. (I updated it today. Use the 13.x source no matter your ISO is based on FreeBSD 12/13/14...)
2. Prebuild the `geom_ventoy` driver and load it by default in `loader.conf`
3. `geom_ventoy.ko` should be locate at `/boot/kernel/geom_ventoy.ko` inside the ISO file.

Ventoy also need to update for this, so after above steps, you can use the following Ventoy CI release to test your ISO file.
https://github.com/ventoy/Ventoy/actions/runs/1461189757
https://www.ventoy.net/en/doc_github_ci.html

Report here if you have any problem.

**Reply by probonopd** [reply]
Noticed that apparently it does not work when the kernel module is compressed with gzip. The module then loads but does not work, unlike other modules. Is this correct?

**Reply by ventoy** [reply]
Yes.
The `geom_ventoy.ko` can not be compressed.

**Comment by probonopd** [comment]
Actually I spoke too early.

I now get `ventoy map data is invalid, get from resource` messages but still **no** `######## ventoy disk <...> ########` and the system does not boot.

Can it make a difference whether the disk is internal (SATA) or external (USB)?

If you would like to have a try for yourself, here is the ISO:
https://github.com/helloSystem/ISO/releases/tag/experimental-13.0

helloSystem normally hides the kernel messages and shows a graphical splash screen, but you can disable it:
When you see the "Loading kernel..." message press the Backspace key to interrupt it. At the bootloader prompt (`OK`) enter `unset boot_mute` so that you can see the kernel messages, then `boot -v` for verbose mode.

**Comment by ventoy** [comment]
I'm sorry. 
I forgot that VENTOY COMPATIBLE flag can not be used in this case. 
So please remove `VENTOY COMPATIBLE` flag and retry. (don't appear `VENTOY COMPATIBLE` string in publisher so as `NOT VENTOY COMPATIBLE` `NOT YET VENTOY COMPATIBLE`  ....   Ventoy use `strstr` to find the flag string.)

Let's explain that:
`VENTOY COMPATIBLE`  means all the work can be done inside the ISO file and there is no need for Ventoy's help.
But this `grom_ventoy.ko` driver need to be filled with some data by Ventoy.
So we can not use `VENTOY COMPATIBLE` flag here, or Ventoy will not write the map data to the driver field and it will not boot.

**Reply by probonopd** [reply]
Removed the flag. Build 0G69 has no more `VENTOY COMPATIBLE`.

**Reply by ventoy** [reply]
Please Use this CI release: https://github.com/ventoy/Ventoy/actions/runs/1465699577   to test 0G69

**Comment by ventoy** [comment]
The Ventoy CI release has some issue, I will commit the update after I test your ISO file.
How to print the dmesg in the desktip environment? Can I boot to shell?

**Reply by probonopd** [reply]
helloSystem normally boots directly into a graphical desktop, but you can boot into safe mode if you want a shell:
When you see the "Loading kernel..." message press the Backspace key to interrupt it. At the bootloader prompt (`OK`) enter `unset boot_mute` so that you can see the kernel messages, then `boot -s` for safe mode.

**Comment by ventoy** [comment]
Please test:
1. Use this CI release: https://github.com/ventoy/Ventoy/actions/runs/1465699577
2. Remove `VENTOY COMPATIBLE` flag. (don't appear `VENTOY COMPATIBLE` string in publisher so as `NOT VENTOY COMPATIBLE` `NOT YET VENTOY COMPATIBLE` .... Ventoy use `strstr` to find the flag string.)

**Comment by ventoy** [comment, marked as answer]
@probonopd 

Ventoy CI release:
https://github.com/ventoy/Ventoy/actions/runs/1466626280
ISO:
https://github.com/helloSystem/ISO/releases/tag/experimental-13.0

I have tested it and it boot OK (successfully boot into the desktop environment).

**Comment by probonopd** [comment]
Happy to report that thanks to your help [helloSystem](https://hellosystem.github.io/) will be **work with Ventoy** starting with version 0.7.0 once it is released. In the meantime there are experimental builds.

Thank you for making this incredibly useful tool and for your support. :+1:  :1st_place_medal:  :100:

**Comment by grahamperrin** [comment]
@ventoy @probonopd 

<https://lists.freebsd.org/archives/freebsd-hackers/2023-November/002649.html>

**Reply by probonopd** [reply]
helloSystem is no longer shipping the Ventoy kernel module on the ISO, because that turned out to be troublesome. Instead, it now lets Ventoy inject it.

However, this doesn't work on 14.0-RC4 unfortunately.
